portedit.1
.\"-
.\" SPDX-License-Identifier: BSD-2-Clause-FreeBSD
.\"
.\" Copyright (c) 2019 Tobias Kortkamp <tobik@FreeBSD.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd November 2, 2019
.Dt PORTEDIT 1
.Os
.Sh NAME
.Nm portedit
.Nd "edit and refactor FreeBSD Ports Collection Makefiles"
.Sh SYNOPSIS
.Nm
.Cm bump-revision
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Ar Makefile
.Nm
.Cm bump-epoch
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Ar Makefile
.Nm
.Cm get
.Ar variable-regexp
.Op Ar Makefile
.Nm
.Cm merge
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Fl e Ar expr
.Op Ar Makefile
.Nm
.Cm sanitize-append
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Ar Makefile
.Nm
.Cm set-version
.Op Fl iu
.Op Fl w Ar wrapcol
.Ar version
.Op Ar Makefile
.Nm
.Cm unknown-targets
.Op Ar Makefile
.Nm
.Cm unknown-vars
.Op Ar Makefile
.Sh DESCRIPTION
.Nm
is a tool for editing and refactoring
.Fx
Ports Collection Makefiles.
.Pp
If the optional
.Ar Makefile
argument is not given, the Makefile will be read from stdin.
.Pp
The following options are shared between the
.Cm bump-epoch ,
.Cm bump-revision ,
.Cm merge ,
.Cm sanitize-append ,
.Cm set-version
commands and
.Xr portfmt 1 .
.Bl -tag -width indent
.It Fl i
Format
.Ar Makefile
in-place instead of writing the result to stdout.
.It Fl u
Leave variables unsorted.
.It Fl w Ar wrapcol
Sets the wrapping column to
.Ar wrapcol
(default: 80).
This is a goal not a hard value.
It will be ignored for several variables like COMMENT, DISTFILES,
MASTER_SITES, etc.
.Pp
With a value of -1 it is ignored for all variables.
Variables with wrapped tokens over multiple lines will be concatenated
onto a single line.
.El
.Sh COMMANDS
.Bl -tag -width 2n
.It Xo
.Nm
.Cm bump-epoch
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Ar Makefile
.Xc
.Pp
Same as
.Cm bump-revision
but for PORTEPOCH.
.It Xo
.Nm
.Cm bump-revision
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Ar Makefile
.Xc
.Pp
Increments PORTREVISION by one.
It will be inserted if not already present in the Makefile.
If PORTREVISION is not currently in the right place it will be moved
as part of the operation.
.It Xo
.Nm
.Cm get
.Ar variable-regexp
.Op Ar Makefile
.Xc
.Pp
Returns the raw tokens of a variable matching the extended regular
expression
.Po see
.Xr re_format 7 Pc
.Ar variable-regexp .
.It Xo
.Nm
.Cm merge
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Fl e Ar expr
.Op Ar Makefile
.Xc
.Pp
Merges files in
.Xr make 1
syntax into
.Ar Makefile .
The input is read from stdin in the normal case, but
one or more
.Fl e
can be used to specify the input instead.
Each
.Fl e
is treated as if it would be an input line.
.Pp
This is currently limited to variables only.
Conditionals and targets are ignored.
.Pp
The
.Sy !=
modifier is overloaded to delete variables.
.Sy +=
will append to the specified variable.
No deduplication of tokens happens yet.
.Pp
Variables in the target Makefile should be in the order as suggested
by
.Xr portclippy 1
for best results when inserting new variables.
.It Xo
.Nm
.Cm sanitize-append
.Op Fl iu
.Op Fl w Ar wrapcol
.Op Ar Makefile
.Xc
.Pp
Sanitize
.Sy +=
in Makefiles before
.Sy bsd.port.options.mk
or
.Sy bsd.port.pre.mk .
This prevents leaking variable values from the build environment.
CXXFLAGS, CFLAGS, LDFLAGS, RUSTFLAGS will not be sanitized since
users might want to set optimization flags that affect port builds
in their environment.
.It Xo
.Nm
.Cm set-version
.Op Fl iu
.Op Fl w Ar wrapcol
.Ar version
.Op Ar Makefile
.Xc
.Pp
Sets the port version.
If the port currently has PORTVERSION it will be used again for the
new version.
Otherwise it defaults to DISTVERSION.
.Pp
PORTREVISION will be removed (reset to 0).
However special care is taken for slave ports where PORTREVISION
will not be removed and will be set to 0 explicitly to prevent
inheriting the revision from the master port.
.Pp
It tries to be smart based on the format of
.Ar version .
.Pp
If
.Ar version
starts with the current value of DISTVERSIONPREFIX or ends with the
current value of DISTVERSIONSUFFIX, it will be trimmed accordingly
to set DISTVERSION or PORTVERSION correctly.
.Pp
If
.Ar version
is in the format as returned by
.Cm git describe --tags ,
DISTVERSION{PREFIX,SUFFIX} will be set accordingly.
.It Xo
.Nm
.Cm unknown-targets
.Op Ar Makefile
.Xc
.Pp
List targets that are not known to
.Nm
or
.Xr portclippy 1 .
.It Xo
.Nm
.Cm unknown-vars
.Op Ar Makefile
.Xc
.Pp
List variables that are not known to
.Nm
or
.Xr portclippy 1 .
.El
.Sh EXAMPLES
Append
.Sy pkgconfig
to USES and delete MASTER_SITES:
.Bd -literal -offset indent
$ printf 'USES+=pkgconfig\\nMASTER_SITES!=' | portedit merge -i Makefile
.Ed
.Pp
Or alternatively using
.Fl e :
.Bd -literal -offset indent
$ portedit merge -i -e 'USES+=pkgconfig' -e 'MASTER_SITES!=' Makefile
.Ed
.Pp
During maintainance of USES=cargo ports we have to regenerate
CARGO_CRATES and related variables based on output of
.Cm make cargo-crates .
We can merge the output of this command into our Makefile with
.Nm Cm merge :
.Bd -literal -offset indent
$ make cargo-crates
CARGO_CRATES=	# ...
$ make cargo-crates | portedit merge -i Makefile
.Ed
.Pp
We have to run
.Cm make cargo-crates
twice since the first time the output might not be clean and might
contain framework messages.
.Pp
A port that is using a versioning scheme that is derived from
.Cm git describe --tags
might have
.Bd -literal -offset indent
DISTVERSIONPREFIX=	v
DISTVERSION=	0.9.2-259
DISTVERSIONSUFFIX=	-gc07c115
.Ed
.Pp
To update it we can run
.Bd -literal -offset indent
$ portedit set-version -i v0.9.2-279-gc23f123
.Ed
.Pp
The triple will have been updated accordingly:
.Bd -literal -offset indent
DISTVERSIONPREFIX=	v
DISTVERSION=	0.9.2-279
DISTVERSIONSUFFIX=	-gc23f123
.Ed
.Sh SEE ALSO
.Xr portclippy 1 ,
.Xr portfmt 1
.Sh AUTHORS
.An Tobias Kortkamp Aq Mt tobik@FreeBSD.org


portedit.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
 *
 * Copyright (c) 2019 Tobias Kortkamp <tobik@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "config.h"

#if HAVE_ERR
# include <err.h>
#endif
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <unistd.h>

#include "array.h"
#include "mainutils.h"
#include "parser.h"
#include "util.h"

static int bump_epoch(struct ParserSettings *, int, char *[]);
static int bump_revision(struct ParserSettings *, int, char *[]);
static int get_variable(struct ParserSettings *, int, char *[]);
static int merge(struct ParserSettings *, int, char *[]);
static int sanitize_append(struct ParserSettings *, int, char *[]);
static int set_version(struct ParserSettings *, int, char *[]);
static int unknown_targets(struct ParserSettings *, int, char *[]);
static int unknown_vars(struct ParserSettings *, int, char *[]);
static void bump_epoch_usage(void);
static void bump_revision_usage(void);
static void get_variable_usage(void);
static void merge_usage(void);
static void sanitize_append_usage(void);
static void set_version_usage(void);
static void unknown_targets_usage(void);
static void unknown_vars_usage(void);
static void usage(void);

static struct Parser *read_file(struct ParserSettings *, FILE **, FILE **, int *, char **[], int);

struct PorteditCommand {
	const char *name;
	int (*main)(struct ParserSettings *, int, char *[]);
};

static struct PorteditCommand cmds[] = {
	{ "bump-epoch", bump_epoch },
	{ "bump-revision", bump_revision },
	{ "get", get_variable },
	{ "merge", merge },
	{ "unknown-targets", unknown_targets },
	{ "unknown-vars", unknown_vars },
	{ "sanitize-append", sanitize_append },
	{ "set-version", set_version },
};

int
bump_epoch(struct ParserSettings *settings, int argc, char *argv[])
{
	if (argc < 2) {
		bump_epoch_usage();
	}
	argv++;
	argc--;

	if (!read_common_args(&argc, &argv, settings, "iuw:", NULL)) {
		bump_epoch_usage();
	}

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 0);
	if (parser == NULL) {
		bump_epoch_usage();
	}

	int error = parser_edit(parser, edit_bump_revision, "PORTEPOCH");
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	return 0;
}

int
bump_revision(struct ParserSettings *settings, int argc, char *argv[])
{
	if (argc < 2) {
		bump_revision_usage();
	}
	argv++;
	argc--;

	if (!read_common_args(&argc, &argv, settings, "iuw:", NULL)) {
		bump_revision_usage();
	}

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 0);
	if (parser == NULL) {
		bump_revision_usage();
	}

	int error = parser_edit(parser, edit_bump_revision, NULL);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	return 0;
}

int
get_variable(struct ParserSettings *settings, int argc, char *argv[])
{
	settings->behavior |= PARSER_OUTPUT_RAWLINES;

	if (argc < 3) { 
		get_variable_usage();
	}
	const char *var = argv[2];
	argv += 3;
	argc -= 3;

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 0);
	if (parser == NULL) {
		get_variable_usage();
	}

	int error = parser_edit(parser, edit_output_variable_value, var);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	return 0;
}

int
merge(struct ParserSettings *settings, int argc, char *argv[])
{
	settings->behavior |= PARSER_DYNAMIC_PORT_OPTIONS;

	if (argc < 2) {
		merge_usage();
	}
	argv++;
	argc--;

	struct Array *expressions = array_new(sizeof(char *));
	if (!read_common_args(&argc, &argv, settings, "e:iuw:", expressions)) {
		merge_usage();
	}
	if (argc == 0) {
		merge_usage();
	}

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 1);
	if (parser == NULL) {
		merge_usage();
	}

	struct Parser *subparser = parser_new(settings);
	int error = PARSER_ERROR_OK;
	if (array_len(expressions) > 0) {
		for (size_t i = 0; i < array_len(expressions); i++) {
			char *expr = array_get(expressions, i);
			error = parser_read_from_buffer(subparser, expr, strlen(expr));
			if (error != PARSER_ERROR_OK) {
				errx(1, "%s", parser_error_tostring(subparser));
			}
			free(expr);
		}
	} else {
		error = parser_read_from_file(subparser, stdin);
		if (error != PARSER_ERROR_OK) {
			errx(1, "%s", parser_error_tostring(subparser));
		}
	}
	array_free(expressions);
	expressions = NULL;

	error = parser_read_finish(subparser);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(subparser));
	}
	error = parser_merge(parser, subparser, PARSER_MERGE_SHELL_IS_DELETE | PARSER_MERGE_COMMENTS);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(subparser);

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	return 0;
}

int
sanitize_append(struct ParserSettings *settings, int argc, char *argv[])
{
	settings->behavior |= PARSER_SANITIZE_APPEND;

	if (argc < 1) {
		sanitize_append_usage();
	}
	argv++;
	argc--;

	if (!read_common_args(&argc, &argv, settings, "iuw:", NULL)) {
		sanitize_append_usage();
	}

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 1);
	if (parser == NULL) {
		sanitize_append_usage();
	}

	int error = parser_edit(parser, refactor_sanitize_append_modifier, NULL);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	return 0;
}

int
set_version(struct ParserSettings *settings, int argc, char *argv[])
{
	if (argc < 2) {
		set_version_usage();
	}
	argv++;
	argc--;

	if (!read_common_args(&argc, &argv, settings, "iuw:", NULL)) {
		set_version_usage();
	}

	if (argc < 1) { 
		set_version_usage();
	}
	const char *version = argv[0];
	argv++;
	argc--;

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 1);
	if (parser == NULL) {
		set_version_usage();
	}

	int error = parser_edit(parser, edit_set_version, version);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	return 0;
}

int
unknown_targets(struct ParserSettings *settings, int argc, char *argv[])
{
	settings->behavior |= PARSER_OUTPUT_RAWLINES;

	argv += 2;
	argc -= 2;

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 0);
	if (parser == NULL) {
		unknown_targets_usage();
	}

	struct Array *unknowns = NULL;
	enum ParserError error = parser_edit(parser, edit_output_unknown_targets, &unknowns);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	if (array_len(unknowns) > 0) {
		array_free(unknowns);
		return 1;
	}
	array_free(unknowns);

	return 0;
}

int
unknown_vars(struct ParserSettings *settings, int argc, char *argv[])
{
	settings->behavior |= PARSER_OUTPUT_RAWLINES;

	argv += 2;
	argc -= 2;

	FILE *fp_in = stdin;
	FILE *fp_out = stdout;
	struct Parser *parser = read_file(settings, &fp_in, &fp_out, &argc, &argv, 0);
	if (parser == NULL) {
		unknown_vars_usage();
	}

	struct Array *unknowns = NULL;
	enum ParserError error = parser_edit(parser, edit_output_unknown_variables, &unknowns);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	error = parser_output_write_to_file(parser, fp_out);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	parser_free(parser);

	fclose(fp_out);
	if (fp_out != fp_in) {
		fclose(fp_in);
	}

	if (array_len(unknowns) > 0) {
		array_free(unknowns);
		return 1;
	}
	array_free(unknowns);

	return 0;
}

void
bump_epoch_usage()
{
	fprintf(stderr, "usage: portedit bump-epoch [-iu] [-w wrapcol] [Makefile]\n");
	exit(EX_USAGE);
}

void
bump_revision_usage()
{
	fprintf(stderr, "usage: portedit bump-revision [-iu] [-w wrapcol] [Makefile]\n");
	exit(EX_USAGE);
}

void
get_variable_usage()
{
	fprintf(stderr, "usage: portedit get <variable-regexp> [Makefile]\n");
	exit(EX_USAGE);
}

void
merge_usage()
{
	fprintf(stderr, "usage: portedit merge [-iu] [-w wrapcol] [-e expr] [Makefile]\n");
	exit(EX_USAGE);
}

void
sanitize_append_usage()
{
	fprintf(stderr, "usage: portedit sanitize-append [-iu] [-w wrapcol] [Makefile]\n");
	exit(EX_USAGE);
}

void
set_version_usage()
{
	fprintf(stderr, "usage: portedit set-version [-iu] [-w wrapcol] <version> [Makefile]\n");
	exit(EX_USAGE);
}

void
unknown_targets_usage()
{
	fprintf(stderr, "usage: portedit unknown-targets [Makefile]\n");
	exit(EX_USAGE);
}

void
unknown_vars_usage()
{
	fprintf(stderr, "usage: portedit unknown-vars [Makefile]\n");
	exit(EX_USAGE);
}

void
usage()
{
	fprintf(stderr, "usage: portedit <command> [<args>]\n\n");
	fprintf(stderr, "Supported commands:\n");
	fprintf(stderr, "\t%-16s%s\n", "bump-epoch", "Bump and sanitize PORTEPOCH");
	fprintf(stderr, "\t%-16s%s\n", "bump-revision", "Bump and sanitize PORTREVISION");
	fprintf(stderr, "\t%-16s%s\n", "get", "Get raw variable tokens");
	fprintf(stderr, "\t%-16s%s\n", "merge", "Merge variables into the Makefile");
	fprintf(stderr, "\t%-16s%s\n", "sanitize-append", "Sanitize += before bsd.port.{options,pre}.mk");
	fprintf(stderr, "\t%-16s%s\n", "set-version", "Bump port version, set DISTVERSION{,PREFIX,SUFFIX}");
	fprintf(stderr, "\t%-16s%s\n", "unknown-targets", "List unknown targets");
	fprintf(stderr, "\t%-16s%s\n", "unknown-vars", "List unknown variables");
	exit(EX_USAGE);
}

struct Parser *
read_file(struct ParserSettings *settings, FILE **fp_in, FILE **fp_out, int *argc, char **argv[], int keep_stdin_open)
{
	if (!open_file(argc, argv, settings, fp_in, fp_out, keep_stdin_open)) {
		if (*fp_in == NULL) {
			err(1, "fopen");
		} else {
			return NULL;
		}
	}
	if (!can_use_colors(*fp_out)) {
		settings->behavior |= PARSER_OUTPUT_NO_COLOR;
	}

	enter_sandbox();

	struct Parser *parser = parser_new(settings);
	enum ParserError error = parser_read_from_file(parser, *fp_in);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}
	error = parser_read_finish(parser);
	if (error != PARSER_ERROR_OK) {
		errx(1, "%s", parser_error_tostring(parser));
	}

	return parser;
}

int
main(int argc, char *argv[])
{
	if (argc < 2) {
		usage();
	}
	const char *command = argv[1];

	struct ParserSettings settings;
	parser_init_settings(&settings);
	settings.behavior = PARSER_COLLAPSE_ADJACENT_VARIABLES | PARSER_DEDUP_TOKENS |
		PARSER_OUTPUT_REFORMAT | PARSER_OUTPUT_EDITED |
		PARSER_KEEP_EOL_COMMENTS;

	for (size_t i = 0; i < nitems(cmds); i++) {
		if (strcmp(command, cmds[i].name) == 0) {
			return cmds[i].main(&settings, argc, argv);
		}
	}

	usage();
	abort();
}
